File: AlgorithmsPackJavaSolutions/Number of Operations to Make Network Connected.java
class Solution {
    private class UnionFind {
        private int[] rank;
        private int[] parent;
        public int forests;

        public UnionFind(int n) {
            rank = new int[n];
            parent = new int[n];
            forests = n;

            for (int i = 0; i < n; ++i) {
                parent[i] = i;
                rank[i] = 1;
            }
        }

        private void link(int x, int y) {
            if (rank[x] > rank[y]) {
                int temp = x;
                x = y;
                y = temp;
            }

            parent[x] = y;
            if (rank[x] == rank[y])
                rank[y]++;
        }

        private int find_set(int x) {
            if (x != parent[x])
                parent[x] = find_set(parent[x]);
            return parent[x];
        }

        public boolean union_sets(int x, int y) {
            x = find_set(x);
            y = find_set(y);
            if (x != y) {
                link(x, y);
                forests--;
            }
            return x != y;
        }
    }

    public int makeConnected(int n, int[][] connections) {
        if (connections.length < n - 1)
            return -1;

        UnionFind uf = new UnionFind(n);
        for (int[] edge : connections)
            uf.union_sets(edge[0], edge[1]);

        return uf.forests - 1;
    }
}

File: AlgorithmsPackPythonSolutions/Number of Operations to Make Network Connected.py
class Solution:
    class UnionFind:
        def __init__(self, n):
            self.rank = [1] * n
            self.parent = list(range(n))
            self.forests = n

        def link(self, x, y):
            if self.rank[x] > self.rank[y]:
                x, y = y, x
            self.parent[x] = y
            if self.rank[x] == self.rank[y]:
                self.rank[y] += 1

        def find_set(self, x):
            if x != self.parent[x]:
                self.parent[x] = self.find_set(self.parent[x])
            return self.parent[x]

        def union_sets(self, x, y):
            x, y = self.find_set(x), self.find_set(y)
            if x != y:
                self.link(x, y)
                self.forests -= 1
            return x != y

    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        if len(connections) < n - 1:
            return -1

        uf = self.UnionFind(n)
        for edge in connections:
            uf.union_sets(edge[0], edge[1])

        return uf.forests - 1
